Neo4j.Driver/Neo4j.Driver/Internal/IO/ChunkReader.cs

```diff
@@ -0,0 +1,236 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Neo4j.Driver.V1;
+
+namespace Neo4j.Driver.Internal.IO
+{
+    internal class ChunkReader: IChunkReader
+    {
+        private readonly Stream _downStream;
+        private readonly MemoryStream _chunkStream;
+        private readonly ILogger _logger;
+        private long _lastWritePosition = 0;
+        private long _lastReadPosition = 0;
+
+        private readonly byte[] _chunkSizeBuffer = new byte[2];
+        private readonly byte[] _buffer = new byte[8 * 1024];
+        private int _currentChunkSize = -1;
+
+        public ChunkReader(Stream downStream)
+            : this(downStream, null)
+        {
+            
+        }
+
+        public ChunkReader(Stream downStream, ILogger logger)
+            : this(downStream, new MemoryStream(), logger)
+        {
+            
+        }
+
+        internal ChunkReader(Stream downStream, MemoryStream chunkStream, ILogger logger)
+        {
+            Throw.ArgumentNullException.IfNull(downStream, nameof(downStream));
+            Throw.ArgumentOutOfRangeException.IfFalse(downStream.CanRead, nameof(downStream));
+
+            Throw.ArgumentNullException.IfNull(chunkStream, nameof(chunkStream));
+
+            _downStream = downStream;
+            _chunkStream = chunkStream;
+            _logger = logger;
+        }
+
+        public void ReadNextChunk(Stream targetStream)
+        {
+            while (true)
+            {
+                // We have not received the chunk size yet, but it can be read from the buffered data.
+                if (_currentChunkSize == -1 && HasBytesAvailable(_chunkSizeBuffer.Length))
+                {
+                    ReadFromChunkStream(_chunkSizeBuffer, 0, _chunkSizeBuffer.Length);
+
+                    _currentChunkSize = PackStreamBitConverter.ToUInt16(_chunkSizeBuffer);
+                    if (_currentChunkSize == 0)
+                    {
+                        break;
```

Comment:
So this while loop will not stop until we've reached a `0x00 00`, does this mean that we will not stop putting more data into `targetStream` (a.k.a. `boltReader._bufferStream`) until we pulled the whole message into `targetStream`?

As I committed in `boltReader` I do not think we could assume that we could always pull the whole message into memory. CC @technige I need some inputs from you too.

Similar for writer, when a chunk is finished, then we should flush it immediately.